File: leetcode-solutions/ruby/704. Binary Search.rb
# https://leetcode.com/problems/binary-search/

# @param {Integer[]} nums
# @param {Integer} target
# @return {Integer}
def search(nums, target)
  return -1 if nums.size.zero?
  left = 0
  right = nums.size - 1
  
  while right >= left
      mid = (left + right) / 2
      
      if nums[mid] == target
          return mid
      elsif target > nums[mid]
          left = mid + 1
      else
          right = mid - 1
      end
  end

  -1
end

File: leetcode-solutions/python/Algorithm-1/704. Binary Search.py
# https://leetcode.com/problems/binary-search/

class Solution:
    def search(self, nums: List[int], target: int) -> int:
        s, e = 0, len(nums) - 1
        
        while s <= e:
            mid = (e - s) // 2 + s
            if nums[mid] == target:
                return mid
            elif nums[mid] > target:
                e = mid - 1
            else:
                s = mid + 1
        return -1
